File: app/src/main/java/com/felipecsl/gifimageview/app/Blur.java
package com.felipecsl.gifimageview.app;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.v8.renderscript.Allocation;
import android.support.v8.renderscript.Element;
import android.support.v8.renderscript.RenderScript;
import android.support.v8.renderscript.ScriptIntrinsicBlur;

public class Blur {
  private static final float BLUR_RADIUS = 25f;

  private final RenderScript rs;
  private ScriptIntrinsicBlur script;
  private Allocation input;
  private Allocation output;
  private boolean configured = false;
  private Bitmap tmp;
  private int[] pixels;

  public static Blur newInstance(Context context) {
    return new Blur(context);
  }

  private Blur(Context context) {
    rs = RenderScript.create(context);
  }

  public Bitmap blur(Bitmap image) {
    if (image == null)
      return null;

    image = RGB565toARGB888(image);
    if (!configured) {
      input = Allocation.createFromBitmap(rs, image);
      output = Allocation.createTyped(rs, input.getType());
      script = ScriptIntrinsicBlur.create(rs, Element.U8_4(rs));
      script.setRadius(BLUR_RADIUS);
      configured = true;
    } else
      input.copyFrom(image);

    script.setInput(input);
    script.forEach(output);
    output.copyTo(image);

    return image;
  }

  private Bitmap RGB565toARGB888(Bitmap img) {
    int numPixels = img.getWidth() * img.getHeight();

    //Create a Bitmap of the appropriate format.
    if (tmp == null) {
      tmp = Bitmap.createBitmap(img.getWidth(), img.getHeight(), Bitmap.Config.ARGB_8888);
      pixels = new int[numPixels];
    }

    //Get JPEG pixels.  Each int is the color values for one pixel.
    img.getPixels(pixels, 0, img.getWidth(), 0, 0, img.getWidth(), img.getHeight());

    //Set RGB pixels.
    tmp.setPixels(pixels, 0, tmp.getWidth(), 0, 0, tmp.getWidth(), tmp.getHeight());

    return tmp;
  }
}

File: app/src/main/java/com/felipecsl/gifimageview/app/Blur.kt
package com.felipecsl.gifimageview.app

import android.content.Context
import android.graphics.Bitmap
import androidx.renderscript.Allocation
import androidx.renderscript.Element
import androidx.renderscript.RenderScript
import androidx.renderscript.ScriptIntrinsicBlur

class Blur private constructor(context: Context) {
    private val rs: RenderScript
    private var script: ScriptIntrinsicBlur? = null
    private var input: Allocation? = null
    private var output: Allocation? = null
    private var configured = false
    private var tmp: Bitmap? = null
    private lateinit var pixels: IntArray
    fun blur(image: Bitmap?): Bitmap? {
        var bitmapImage: Bitmap? = image ?: return null
        bitmapImage = bitmapImage?.let { RGB565toARGB888(it) }
        if (!configured) {
            input = Allocation.createFromBitmap(rs, bitmapImage)
            output = Allocation.createTyped(rs, input?.type)
            script = ScriptIntrinsicBlur.create(rs, Element.U8_4(rs))
            script?.setRadius(BLUR_RADIUS)
            configured = true
        } else input?.copyFrom(bitmapImage)
        script?.setInput(input)
        script?.forEach(output)
        output?.copyTo(bitmapImage)
        return bitmapImage
    }

    private fun RGB565toARGB888(img: Bitmap): Bitmap? {
        val numPixels = img.width * img.height

        //Create a Bitmap of the appropriate format.
        if (tmp == null) {
            tmp = Bitmap.createBitmap(img.width, img.height, Bitmap.Config.ARGB_8888)
            pixels = IntArray(numPixels)
        }

        //Get JPEG pixels.  Each int is the color values for one pixel.
        img.getPixels(pixels, 0, img.width, 0, 0, img.width, img.height)

        //Set RGB pixels.
        tmp?.setPixels(pixels, 0, tmp!!.width, 0, 0, tmp!!.width, tmp!!.height)
        return tmp
    }

    companion object {
        private const val BLUR_RADIUS = 25f
        @JvmStatic
        fun newInstance(context: Context): Blur {
            return Blur(context)
        }
    }

    init {
        rs = RenderScript.create(context)
    }
}